12773 stiebeleltron enhancements

tb4jc opened the pull request:
<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").

ATTENTION: Don't use "git merge" when working with your pull request branch!
This can clutter your Git history and make your PR unusable.
Use "git rebase" instead. See this forum post for further details:
https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358

All PRs should be created using the "main" branch as base.
Important bugfixes are cherry-picked by maintainers to the patch release branch after a PR has been merged.

Add one or more appropriate labels to make your PR show up in the release notes.
E.g. enhancement, bug, documentation, new binding
This can only be done by yourself if you already contributed to this repo.

If your PR's code is not backward compatible with previous releases (which
should be avoided), add a message to the release notes by filing another PR:
https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/bin/update.lst

# Title

Provide a short summary in the *Title* above. It will show up in the release notes.
For example:
- [homematic] Improve communication with weak signal devices
- [timemachine][WIP] Initial contribution

# Description

Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/bindings/#include-the-binding-in-the-build
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work

# Testing

Your pull request will automatically be built and available under the following folder:
https://openhab.jfrog.io/ui/native/libs-pullrequest-local/org/openhab/addons/bundles/

It is a good practice to add a URL to your built JAR in this pull request description,
so it is easier for the community to test your Add-on.
If your pull request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Marketplace:
https://community.openhab.org/c/marketplace/69

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->
# Description
This is a enhancement Pull Request for the Stiebel Eltron ISG binding. It is releated to the issue https://github.com/openhab/openhab-addons/issues/12773.
- I added three new Things for the different WPMs "WPMsystem", "WPM3" and "WPM3i", one thing to support the Energy Management settings and information (SG Ready - only supported with an ISG plus)
- The new Things support the retrieval and control of more items of the heat pump.
- The binding still contains the original Thing to stay backward compatible.
- I added a German translation file.

Check the new README.md for more information about the new Things and channels.

# Testing
I have tested the enhancement only with version 4.2.3 and using it since about 3 years with earlier releases.
The jar file for Release 4.2.3 and a snapshot build for 5.0.0 are [here](https://1drv.ms/f/s!AsoLHChjI3wLtxG7UDVr_qOrDtEr?e=eWz7If) available. Note: I didn't tested the 5.0.0 snapshot build. I don't have an 5.0.0 beta installation.

GerdZanker commented:
Hello,

I tested the PR code with my 5.0.0-SNAPSHOT installation. Creating a bridge thing and an ISG + WPMSystem worked for me. I can see "old" value and get new values from the my real heating system. Good work @tb4jc !

Here some feedback for improvement & discussion:
1. I miss the _backflow_ temperature of my heating system. It seems to be communicated with modbus register 542. Therefore I suggest to add the Heatpump1 modbus registers 542-548. The value of register 516 is always shown as 0 in my case.
2. I took a look at the code and can see several duplicated code parts inside the new files. Is a refactoring required before a PR merge?
3. The trace statements give details. But the output could be improved and better structured to really trace down what is happening inside the bundle. What do you think?
4. Perhaps the ISG can be discovered in the network and the ISG/WPM thing can be discovered afterwards based on try and error/success - but this an improvement idea if this PR is finished.

tb4jc commented:
Hi Gerd

I'm sorry I didn't answer earlier. I'm ill since Silvester, now it's the first time at startet my computer again since then.

Thank you very much for your feedback. I appreciate it.

1. Have you checked the values directly on the ISG with a Modbus tool (e.g. I'm using https://sourceforge.net/projects/qmodmaster)? At least in the code, I haven't find any differences regarding parsing and handling of the value of 'ACTUAL RETURN TEMPERATURE'.
Besides, yes I know there are many 'system values' missing for WPMsystem and WPM3 (all Heat Pump 1 to 6 and different circuits). If you can approve that adding at least Heat Pump 1 would be enough for you, then I could surely add it.
2. Yes, I fully agree, there is a lot of duplicated code. But as I'm not that used to Java and all its features, I made it as simple as possible; besides I don't really have another idea right how it could be simplified. My goal now also was to have the enhancements in the main branch as soon as possible, so that everybody can use it in the future.
Who is charge to decide if that needs to be done before merging? I don't know.
3. I didn't really care that much about the existing tracing code. I exteneded it where I needed to verify read values, but else it's more or less the same as before.
4. I don't know how I could add this feature. But I think it wouldn't really be such a big help as the required info can easily be found: IP address of ISG/ISG+, WPM version on diagnostic system page: 
![image](https://github.com/user-attachments/assets/3acab117-2be3-42cf-b9c9-5102ac5f99b8)

Does anybody know how we can proceed, if @pail23 is not going to review my changes? (it looks like he moved on th Home Assistant).

GerdZanker commented:
> I'm sorry I didn't answer earlier. I'm ill since Silvester, now it's the first time at startet my computer again since then.

No problem at all and good that you well again. My own response time is more like a week not not within a day.


>     Have you checked the values directly on the ISG with a Modbus tool (e.g. I'm using https://sourceforge.net/projects/qmodmaster)? 

No, not yet. But I will try the modbus tool and check the raw values. But I guess that not all "return temperature" registered are filled.


Regarding

1. I think the additional values for Heatpump1 are sufficient. I already modified your code and will proved a patch soon.

2. OK, fine for me - First goal is to bring enhancements merged to `main`! I will collect possible improvements and create an issues with my recommended changes after this PR is merge.

3. OK, I use the existing logging output

4. With the Bosch SmartHomeController Bundle I got some experience how the discovery works. Perhaps I can apply it here. I create openhab-addons issue #18067.

GerdZanker commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/java/org/openhab/binding/modbus/stiebeleltron/internal/handler/StiebelEltronHandlerWpm.java:
```diff
@@ -0,0 +1,924 @@
+                getScaled(block.temperatureSourceMin, 10, CELSIUS));
+        updateState(channelUID(GROUP_SYSTEM_INFORMATION_WPM, CHANNEL_SOURCE_PRESSURE),
+                getScaled(block.pressureSource, 100, BAR));
+
```
The last three values of the information block are missing here.

* temperatureHotgas
* pressureHigh
* pressureLow

tb4jc replied:
No, they are not missing as these values are not supported by WPM and WPM3 according to the Stiebel Eltron Modbus TCP documentation.
See list of values in the README.md

GerdZanker commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/resources/OH-INF/thing/heatpump-channel-types.xml:
```diff
@@ -4,139 +4,288 @@
-		<label>Summer Operation Mode</label>
-		<state readOnly="true"/>
+	<channel-type id="heating-pressure-type">
+		<item-type>Number:Pressure</item-type>
```
Default unit for pressure is `pascal`.
But `bar` is usual for heating systems imho.

Shall we change it for all pressures?

tb4jc replied:
Yes, sounds reasonable. The unit mentioned in the document and the ISG web interface is actually 'bar'.
But we should not put it into binding according to the new rules since OH4, but user shall set it via unit metadata.
So, I will update the README.md with the example item configurations to use bar for pressure values.

GerdZanker replied:
I agree, because I want to get the PR merged.
I'm a "default user" and I like good defaults - therefore I would use 'bar' as the default unit and change the channel type:

    <state readOnly="true" pattern="%.2f bar"/>

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/java/org/openhab/binding/modbus/stiebeleltron/internal/parser/SystemInfromationBlockParser.java:
```diff
@@ -17,9 +17,10 @@
+ * @author Thomas Burri - Updated renamed members and fixed typo
  *
  */
 @NonNullByDefault
```
done

tb4jc replied:
Hm, now with the changed class name, we have a merge conflict with the main repo. Looks like build is not triggered anymore.
How to preceed now?

GerdZanker replied:
I guess the conflict is related to the copyright update in the file header.

The usual way is to fetch upstream changes and do a rebase of your branch on top of latest main branch. Resolve the conflicts and do a forced push of your branch (hope this summary is sufficient, else don't hesitate to ask again)

tb4jc replied:
Now I'm in the shit.
I made it and now I have all commits from main as well.
Maybe something was to much: I also synched my fork to origin. Or I missed the '--force' when pushing?
How can I get rid of them again?

tb4jc commented:
> 1. I think the additional values for Heatpump1 are sufficient. I already modified your code and will proved a patch soon.

Sorry, didn't mention that I also already have some values in the code. No patch needed from your side, just confirmation with Modbus tool that heat pump 1 values are there.

> 4. With the Bosch SmartHomeController Bundle I got some experience how the discovery works. Perhaps I can apply it here. I create openhab-addons issue [[modbus.stiebeleltron]  Support discovery for IGM and WPM things #18067](https://github.com/openhab/openhab-addons/issues/18067).

Ok, thanks.

GerdZanker commented:
I can confirm to get values for heatpump 1. But I have doubts about the address - need to cross check my results...
![image](https://github.com/user-attachments/assets/2fc9a430-23f4-40c8-a12e-52f5c3ac291a)

tb4jc commented:
Hi Gerd

Ok, I managed somehow to get rid of upstream commits but unfortunately, the review requests are still there.

I found out why you get 0 for 516. Your picture above helped to identify that the ISG value is 36864 (0x9000), which means in other registers OFF (in this case we can interpret it also as invalid I would say). The handler checks for this value and fills in 0.

GerdZanker commented:
>  No patch needed from your side, just confirmation with Modbus tool that heat pump 1 values are there.

Hello @tb4jc, can you please tell be the current state of the code and what changes are planned from your side?
E.g. I assume that the heat pump 1 values are currently pending ...

tb4jc commented:
Hi Gerd
Please check my new build available [here](https://1drv.ms/f/s!AsoLHChjI3wL52tcUGj5yb8hW9q6?e=taUy37.)
tbj4c

tb4jc commented:
@lsiepel @jlaur @lolodomo 
Hi
I'm a little bit in the mess with this PR. As I'm not so used to git I had a commit in the branch with all upstream changes, which lead to the adding of all the review requests. I could remove the commit again somehow, but the review requests stayed.
Any chance to remove them?
The code owner is @pail23 but it looks like he won't review the code as he moved to HA.
How can we proceed now on this PR that it gets a chance to be merged?
Thanks for your support.
Cheers Thomas

tb4jc commented:
Hi

@GerdZanker Any update on your testing?

@pail23 Could you please review the PR? Or is there any other option how this PR can get proceeded?

Cheers Thomas

GerdZanker commented:
Hello @tb4jc, 

> Any update on your testing?

Tests with the latest commit are OK - all seems to work fine and I give my approval.

GerdZanker commented:
I want to ask you @tb4jc regarding older comments from you and me, to clarify what will be merged:

>> I think the additional values for Heatpump1 are sufficient. I already modified your code and will proved a patch soon.

>Sorry, didn't mention that I also already have some values in the code. No patch needed from your side, just confirmation with Modbus tool that heat pump 1 values are there.

With the modbus tool and my own code I see that the heat pump 1 values exist. I am looking forward to get this values, especially the backflow temperator.
These heat pump 1 values are not part of current code.
What is you plan @tb4jc? Will you update your code?

tb4jc commented:
> I want to ask you @tb4jc regarding older comments from you and me, to clarify what will be merged:
> 
> > > I think the additional values for Heatpump1 are sufficient. I already modified your code and will proved a patch soon.
> 
> > Sorry, didn't mention that I also already have some values in the code. No patch needed from your side, just confirmation with Modbus tool that heat pump 1 values are there.
> 
> With the modbus tool and my own code I see that the heat pump 1 values exist. I am looking forward to get this values, especially the backflow temperator. These heat pump 1 values are not part of current code. What is you plan @tb4jc? Will you update your code?

Hi Gerd
Ok, there was a missunderstanding. I only shared my local build with the changes (added HeatPump1) that you could check it before I comit it.
Thomas

tb4jc commented:
Hi Gerd
Changes are committed. Now we have to wait for the review(s).
Cheers
Thomas

GerdZanker commented on bundles/org.openhab.binding.modbus.stiebeleltron/README.md:
```diff
@@ -78,81 +184,181 @@ This group contains general operational information about the device.
+| hotgas-temperature                        | Number:Temperature        | true      | The current hot gas temperature                                      |     |      |   x   |
+| high-pressure                             | Number:Pressure           | true      | The current high pressure                                            |     |      |   x   |
+| low-pressure                              | Number:Pressure           | true      | The current low pressure                                             |     |      |   x   |
+
```
The new heat pump channels could be listed here to be complete

tb4jc replied:
Ok, done.
BTW, you see the power producion and consumption values in the Energy Information Group?
I ask because I have seen that there are more registeres for HP1 and even the NHZ1/2 (3523-3548).
Maybe I have time tomorrow and will add them as well to match HP1 adding in System Information group.

tb4jc replied:
Hi Gerd
Please check latest snapshot build [here ](https://1drv.ms/f/s!AsoLHChjI3wL6FlP3Jpl3kc_uEnb?e=Z8Zl5X)
I didn't commit changes yet.
Cheers Thomas

GerdZanker replied:
> Please check latest snapshot build here. I added the these registers to the Energy & Runtime Block of WPMsystems / WPM3.

With your compiled jar I see values for all Energy Data. All are shown as [kWh] and amount is the same as in IGM web pages (somethings the IGM show a higher precision).

All runtime values are NULL - thing, channel and items are looking good, no exception detected in log file.

tb4jc replied:
Hi Gerd
I have fixed the code (wrong group id used).
You find the latest build [here](https://1drv.ms/f/s!AsoLHChjI3wL6GKfQdbqxLSdE-Cc?e=tonOVd).
Cheers
Thomas

GerdZanker replied:
Hi Thomas,
thank you for the fast update. Now I see three runtime values (Laufzeit Nachheizstufe .... 3546-3548). For all other runtime values I see 117964800s, because I get the "not supported" value of 0x8000.
Compare with 
![image](https://github.com/user-attachments/assets/3bc47060-6a08-4f95-8551-642c153032c9)

I have no idea, why my IGM doesn't return all runtime values, but it is not a problem of the binding code. Therefore I think you are done with the improvements.

tb4jc replied:
Hi Gerd
Thanks for the feedback.
I think in that case I will add a check to all these registers. If the value is 0x8000, I will set it to 0 and make a log entry that the value is not supported. Or is it possible to set a channel to unavailable? That would be the better approach.
Which ISG do you have, normal or ISG+? And which WPM (maybe you have the type in register 5002?) It would also be interesting to have a dump of all your registers with QModMaster.
Thanks for your support.
Cheers
Thomas

GerdZanker replied:
Hello @tb4jc ,
I'm very sorry for my (too) late response, but I wasn't able to find enough time between all the other important things during the last weeks.

Here the requested data
* It's a ISG+ v12.1.0
* the WPM type is 449

And here a screen dump of the registers with QModMaster. Please let me know if there is a possibility to do a register dump into text file which I wasn't to find today...
![QModMaster Reg 0500](https://github.com/user-attachments/assets/61c45dfe-6387-46e8-a288-04714f6fde74) ![QModMaster Reg 1500](https://github.com/user-attachments/assets/c4902f75-46bb-40ca-8e50-82136915f345) ![QModMaster Reg 2500](https://github.com/user-attachments/assets/61bc8461-aaf2-49dd-b3da-e258e77e0b05) ![QModMaster Reg 3500](https://github.com/user-attachments/assets/6d48221c-b7f4-40a0-9508-61a1b37e19e5) ![QModMaster Reg 4000](https://github.com/user-attachments/assets/95fadd70-2a08-4abc-aa9c-ace4c0e20add) ![QModMaster Reg 5000](https://github.com/user-attachments/assets/270b76e9-435f-423c-8213-dcd9590af6b9)

Please let my know what is missing for you to get this PR reviewed and merged.
Bye, bye
Gerd

GerdZanker commented:
Hello @tb4jc 

Thank you for your effort and I appreciate your current work. Therefore I need to comment regarding

> I didn't commit changes yet.

I'm a software developer and work day by day with git and I still struggle sometimes with it ;) . But still I want to ask you, if you can share code with a git branch please?

With shared code I can do a lot more and support you with debugging and fixing issues (e.g. the NULL runtime values). So even if you need more time because of git branches, it could save time to get the PRs done.
We both could use a different dedicated branch in our forked repositories to "exchange" code or a pull request can be put into "work in progress" until it is finished for review by the committers.

tb4jc commented:
> Hello @tb4jc
> 
> Thank you for your effort and I appreciate your current work. Therefore I need to comment regarding
> 
> > I didn't commit changes yet.
> 
> I'm a software developer and work day by day with git and I still struggle sometimes with it ;) . But still I want to ask you, if you can share code with a git branch please?
> 
> With shared code I can do a lot more and support you with debugging and fixing issues (e.g. the NULL runtime values). So even if you need more time because of git branches, it could save time to get the PRs done. We both could use a different dedicated branch in our forked repositories to "exchange" code or a pull request can be put into "work in progress" until it is finished for review by the committers.

Hi Gerd
I have created a new [branch](https://github.com/tb4jc/openhab-addons/tree/stiebeleltron-enhancements-side-branch) as proposed. Additionally, I added you as collaborator to the repo.
So happe co-working :-)
Cheers
Thomas

lsiepel commented:
Is this PR ready to be reviewed or is it work in progress (if so, please mark it as a draft)

lsiepel commented:
Gentle ping @tb4jc, there is also a conflict to solve.

tb4jc commented:
Hi @lsiepel and @GerdZanker 
I finally managed to fix the merge conflicts in github after some troubles with rebasing and need to make a hard reset locally :-(.
The testing of a local build looks good so far.

But actually, I have started in another branch (origin/stiebeleltron-enhancements-side-branch) a new version with just one additional thing, which I rather like to promote as solution for this PR. This new thing includes even more channels (e.g. support for all possible 6 heatpumps on an WPM Sytsem type); for each channel, it compares the return value with 0x8000; and if so, it "disables" it by setting it hard to NULL.
Another advantage is that is less work for maitenance.
@GerdZanker you have access to the repo and could get the branch for testing and code review if you like.

What do you think? Shall I push the new code to the PR branch?
Cheers

tb4jc commented:
Hi
I share my jar files [here](https://1drv.ms/f/c/0b7c2363281c0bca/Ej5cCMKKdvRDkGF1tNAWhsgBQcUO-npkectdqpIw2a4ZPQ?e=rO952R)
Regards
tb4jc

GerdZanker commented:
Hello @tb4jc, thank you for updating the branch and resolving the conflicts. I saw no errors during a fast test.

Regarding your "side-branch":
1. The code needs some smaller changes, before it compiles, because code checks complaining about your log statements. Example:

                -logger.trace("HP" + (idx + 1) + " Return temperature not available");
                +logger.trace("HP{} Return temperature not available", idx + 1);

3. For testing, I created a "stiebeleltron-heatpump-allwpm" think and I can see a lot of channels/items. For many of them I get NULL values, e.g. also for any "Backflow temperature" (I'm personally interested on the backflow temperature and in the comments above you can see that the value is available on modbus).
4. Providing only a single thing for all heat pumps is good. This makes things simpler for users (e.g. less selection possibilities) and increasing code maintaining is a nice combination. I have today too less time to judge if its good to have so many channels providing no value - perhaps channels can be hidden if modbus value is 0x8000.
5. I will continue with more tests and for example trace down to get my backflow temperature visible.

tb4jc commented:
Hi @GerdZanker 

> 1. The code needs some smaller changes, before it compiles, because code checks complaining about your log statements. Example:
>    ```
>             -logger.trace("HP" + (idx + 1) + " Return temperature not available");
>             +logger.trace("HP{} Return temperature not available", idx + 1);
>    ```
I have updated the logging calls. I have added one where the WPM controller ID configured is updated, when a mismatch with the SG Ready register value exists.

> 2. For testing, I created a "stiebeleltron-heatpump-allwpm" think and I can see a lot of channels/items. For many of them I get NULL values, e.g. also for any "Backflow temperature" (I'm personally interested on the backflow temperature and in the comments above you can see that the value is available on modbus).
You should get all value for the heatpump 1 including 'backlow temp', if you set the configuration parameter 'nrOfHps' of your thing to 1.

> 3. Providing only a single thing for all heat pumps is good. This makes things simpler for users (e.g. less selection possibilities) and increasing code maintaining is a nice combination. I have today too less time to judge if its good to have so many channels providing no value - perhaps channels can be hidden if modbus value is 0x8000.
That's the drawback of having just one new, single heatpump thing. It needs to support all registeres, even if not used by the different WPM types. I have searched for a function to remove a channel from a thing but didn't find any.
Maybe somebody has a clue?
The problem is that the current code does not fetch data from the WPM initalization code.
That would be needed e.g. 
- if we would like to fetch the SG Ready register where the WPM type is stored
- create channels based on WPM type and based on return value when accesing them the first time

Maybe this could be done in a next step and keep for now this new code from the side branch as a first enhancement for the existing binding?

> 4. I will continue with more tests and for example trace down to get my backflow temperature visible.
As mentioned in reply to 2. Set 'nrHofHps' to 1 should work. I could not test it though as I don't have a WPM3 or WPMSystem.

Cheers
Thomas

GerdZanker commented:
Hello @tb4jc, you code update is very welcome and all mvn checks are passing.

It was no obvious to me to manually configure the number of heat pumps.
Yes, setting 'nrHofHps' to works for me and I can see expected values and they are looking good and I found no bad nor missing values up to now.

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/resources/OH-INF/thing/heatpump-channel-types.xml:
```diff
@@ -231,37 +299,659 @@
+	</channel-type>
+	<channel-type id="cp1-hotwater-runtime-type">
+		<item-type>Number:Time</item-type>
+		<label>HeatHot Watering Compressor 1 Runtime</label>
```
Hi @lsiepel 
Do you mean I have to adapt all labels, also the ones from thing-types and channel-groups-types?
For channel-types as see the point; then I probably should add some descriptions.
But for thing-types and channel-group-types that might be difficult to achieve.
Cheers

rayb67 commented:
HI Thomas,
i have a openhab 5.0.0. test installation. I have just installed your new 5.0.0 binding instead of the 4.3.2. First channels I check are the 4 SG Ready object. The did not work.
The older binding 4.3.2 work with OH 5.0.0.
My heat pump is a WPC 7 cool , with WPM 3i 12.2.3.260.

Let me know if I can test something.

Best regards
Ralf

tb4jc commented:
Hi Ralf
Thanks for your comment.
Have you enabled SG Ready register retrieval by setting the new configuration value 'pollSgReady'?
Please check and report back.
Thanks!
Cheer Thomas

tb4jc commented:
Hi all

I have updated the configuration for the new thing by adding new class. Now, new parameters are visible when testing jar via addon folder. You can get build with above link.

Cheers
Thomas

rayb67 commented:
> Hi Ralf @rayb67 Thanks for your comment. Have you enabled SG Ready register retrieval by setting the new configuration value 'pollSgReady'? Like this: `Bridge modbus:tcp:bridge "Stiebel Modbus TCP"[ host="your ip", port=502, id=1 ]` `Thing modbus:stiebeleltron-heatpump-allwpm:stiebelEltronWpmComp "Stiebel Eltron Heat Pump V2" (modbus:tcp:bridge) [ wpmControllerId=391, pollSgReady=true ]`
> 
> Please check and report back. Thanks! Cheer Thomas

Hi Thomas,
i have just checked the SG Ready values with the thing properties you request. 
- openhab is 5.1.0 (Build #4784)
- with  org.openhab.binding.modbus.stiebeleltron-5.0.0-SNAPSHOT-2.jar

SGReady is working as expected in my 4.3 installation. I will change my whole setup with the new / reduced thing setup. I will keep you up todate. Thanks.

lsiepel reviewed (changes requested):
Thank you for providing these enhancements. I have covered all files and  except for the readme i have left just a few comments and those can be applied at many places. Please read them as binding wide suggestions.

lsiepel commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/java/org/openhab/binding/modbus/stiebeleltron/internal/handler/StiebelEltronHandlerAllWpm.java:
```diff
@@ -0,0 +1,1582 @@
+ */
+@NonNullByDefault
+public class StiebelEltronHandlerAllWpm extends BaseThingHandler {
+    public abstract class AbstractBasePoller {
```
Nevermind, by coincedence i review another modbus binding and that has it setup the same. Leave it as is.

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/resources/OH-INF/i18n/stiebeleltron_de.properties:
Do you mean I shall remove the whole file?

tb4jc replied:
I removed it

tb4jc replied:
ok done

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/java/org/openhab/binding/modbus/stiebeleltron/internal/parser/SystemParameterBlockParser.java:
```diff
@@ -20,6 +20,7 @@
  * Parses inverter modbus data into a System Parameter Block
  *
  * @author Paul Frank - Initial contribution
+ * @author Thomas Burri - fixed typo
```
correct, removed comment

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/java/org/openhab/binding/modbus/stiebeleltron/internal/parser/SystemStateBlockParser.java:
```diff
@@ -20,6 +20,7 @@
  * Parses inverter modbus data into a System State Block
  *
  * @author Paul Frank - Initial contribution
+ * @author Thomas Burri - fixed typo
```
Correct, removed comment

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/resources/OH-INF/i18n/stiebeleltron.properties:
```diff
@@ -1,57 +1,344 @@
-# thing types
+#Generated by ResourceBundle Editor (http://essiembre.github.io/eclipse-rbe/)
```
ok, changed back and added section headers again.

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/README.md:
```diff
@@ -1,48 +1,81 @@
+Supported controllers are WPMsystem, WPM3 and WPM3i.
+If the SG Ready polling is enabled using configuration parameter *pollSgReady*, the controller id is retrieved and visible in the appropriate channel.
+If the id is known, it can be set in the configuration parameter *wpmControllerId*.
+The binding code works without correct controller id as it checks the retrieved values and sets unavailable channels fix to NULL.
```
ok changed

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/README.md:
```diff
@@ -1,48 +1,81 @@
+
+| Parameter         | Type    | Required | Default if omitted | Description                                                          |
+| ----------------- | ------- | -------- | ------------------ | -------------------------------------------------------------------- |
+| stateBlockLength  | integer | no       | 3                  | Number of retries before giving up reading from this thing           |
```
fixed

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/README.md:
```diff
@@ -52,107 +85,319 @@ This group contains general operational information about the heat pump.
 | is-pumping       | Contact   | true      | OPEN in case the heat pump is currently in pumping mode       |
 | is-summer        | Contact   | true      | OPEN in case the heat pump is currently in summer mode        |
 
+
+
```
ok

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/README.md:
```diff
@@ -52,107 +85,319 @@ This group contains general operational information about the heat pump.
+| defrost-initiated              | Number    | true      | Defrost Initiated: 0=OFF, 1=INITIATED                                  |  x  |  x   |       |
+| active-error                   | Number    | true      | Active Error Number                                                    |  x  |  x   |   x   |
+
+
```
changed

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/README.md:
```diff
@@ -52,107 +85,319 @@ This group contains general operational information about the heat pump.
+The last block can be available for up to 6 heat pumps. The number of available heat pumps shall be set with the configuration paramaeter *nrOfHps*.
+
+
+### SG Ready - Energy Management Settings
```
changed

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/README.md:
````diff
@@ -52,107 +85,319 @@ This group contains general operational information about the heat pump.
+Bridge modbus:tcp:bridge "Stiebel Modbus TCP"[ host="hostname|ip", port=502, id=1 ]
+Thing modbus:heatpump:stiebelEltron "StiebelEltron" (modbus:tcp:bridge) @"Room" [ ]
 ```
+#### Note
````
changed

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/README.md:
```diff
@@ -52,107 +85,319 @@ This group contains general operational information about the heat pump.
+Number:Energy stiebel_eltron_consumption_heat_today  "Heating power consumption today [%.0f kWh]"        { channel="modbus:heatpump:stiebelEltron:energyInformation#consumption_heat_today" }
+Number:Energy stiebel_eltron_consumption_water_today "Water heating power consumption today  [%.0f kWh]" { channel="modbus:heatpump:stiebelEltron:energyInformation#consumption_water_today" }
+Number:Energy stiebel_eltron_consumption_water_total "Water heating power consumption total [%.3f MWh]"  {channel="modbus:heatpump:stiebelEltron:energyInformation#consumption_water_total"}
 
```
changed

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/README.md:
````diff
@@ -196,3 +441,164 @@ Text label="Heat pumpt" icon="temperature" {
 }
 
 ```
+
````
changed

lsiepel commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/resources/OH-INF/config/config-descriptions.xml:
```diff
@@ -4,22 +4,67 @@
 		<parameter name="refresh" type="integer" min="1" unit="s">
-			<label>Polling Interval</label>
-			<description>Poll interval in seconds. Use zero to disable automatic polling.</description>
+			<label>@text/config.stiebeleltron.modbusconfig.refresh.label</label>
```
Will i don't see the point, as each xml tag is added to the i18n by default. But you are true that there are more bindings with these keys. @jlaur or @lolodomo do you have do you know? When reading the documentation this is not clear to me.

jlaur replied:
Pros:
- No redundancy (texts exist only in the i18n properties file).
- No need to run i18n tool (if approach is used everywhere).

Cons:
- Risk of mistakes.
- Harder to read and review.
- If not used everywhere: Inconsistency.

IMHO it is much safer to use texts in the XML files and use the I18N tool to generate the properties. Just to give you an example that it's not a theoretical risk: #17015

tb4jc replied:
done

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/java/org/openhab/binding/modbus/stiebeleltron/internal/StiebelEltronHandlerFactory.java:
```diff
@@ -31,12 +33,22 @@
 public class StiebelEltronHandlerFactory extends BaseThingHandlerFactory {
 
-    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Set.of(THING_TYPE_HEATPUMP);
+    @SuppressWarnings("serial")
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = new HashSet<>() {
+        {
+            add(THING_TYPE_HEATPUMP);
+            add(THING_TYPE_STIEBELELTRON_HEATPUMP_ALLWPM);
+        }
+    };
```
yes, true.
I changed it for the first approach with 3 additional thing types.
But now with just one additionl, easier like that.

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/java/org/openhab/binding/modbus/stiebeleltron/internal/handler/StiebelEltronHandlerAllWpm.java:
```diff
@@ -0,0 +1,1582 @@
+                throw new StiebelEltronException("Unsupported unit");
+            }
+        }
+        if (command instanceof DecimalType) {
+            DecimalType c = (DecimalType) command;
+            return (short) (c.doubleValue() * scaleFactor);
```
changed

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/java/org/openhab/binding/modbus/stiebeleltron/internal/handler/StiebelEltronHandlerAllWpm.java:
```diff
@@ -0,0 +1,1582 @@
+     * @return short the value of the command as short
+     */
+    private short getInt16Value(Command command) throws StiebelEltronException {
+        if (command instanceof DecimalType) {
+            DecimalType c = (DecimalType) command;
+            return c.shortValue();
```
changed

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/java/org/openhab/binding/modbus/stiebeleltron/internal/handler/StiebelEltronHandlerAllWpm.java:
```diff
@@ -0,0 +1,1582 @@
+                        case CHANNEL_RESTART_ISG:
+                            short restartIsgCmd = getInt16Value(command);
+                            if (restartIsgCmd != CHANNEL_RESTART_ISG_CMD_SERVICE_KEY) {
+                                logger.info("RESTART-ISG command received with value {} - but ignored right now",
```
changed the same

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/java/org/openhab/binding/modbus/stiebeleltron/internal/handler/StiebelEltronHandlerAllWpm.java:
```diff
@@ -0,0 +1,1582 @@
+
+        if (comms == null) {
+            @SuppressWarnings("null")
+            String label = Optional.ofNullable(getBridge()).map(b -> b.getLabel()).orElse("<null>");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE,
+                    String.format("Bridge '%s' not completely initialized", label));
```
Just reused code from previous handler. Could you help with your two points? Suggestions?

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/java/org/openhab/binding/modbus/stiebeleltron/internal/handler/StiebelEltronHandlerAllWpm.java:
```diff
@@ -0,0 +1,1582 @@
+
+        ModbusEndpointThingHandler slaveEndpointThingHandler = getEndpointThingHandler();
+        if (slaveEndpointThingHandler == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE, "Bridge is offline");
```
Removed

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/java/org/openhab/binding/modbus/stiebeleltron/internal/handler/StiebelEltronHandlerAllWpm.java:
```diff
@@ -0,0 +1,1582 @@
+
+import net.wimpi.modbus.Modbus;
+
+/**
+ * The {@link Modbus.StiebelEltronHandler} is responsible for handling commands,
+ * which are sent to one of the channels and for polling the modbus.
+ *
+ * @author Paul Frank - Initial contribution
+ * @author Thomas Burri - Extended thing handler for a WPM compatible heat pump
+ */
```
updated

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/java/org/openhab/binding/modbus/stiebeleltron/internal/handler/StiebelEltronHandlerAllWpm.java:
```diff
@@ -0,0 +1,1582 @@
+     *         and datatype 7 (100) in the stiebel eltron modbus documentation)
+     */
+    private short getScaledInt16Value(Command command, Integer scaleFactor) throws StiebelEltronException {
+        if (command instanceof QuantityType) {
+            QuantityType<?> c = ((QuantityType<?>) command).toUnit(CELSIUS);
```
sorry, yes; I pasted the old code. The new one looks like this (small diff with additional '<?>'):
```
        if (command instanceof QuantityType<?> quantityCommand)
            QuantityType<?> c = quantityCommand.toUnit(CELSIUS);
```
With this, I don't get any complain from Eclipse.

lsiepel replied:
oh yes, perfectly fine!

tb4jc commented on bundles/org.openhab.binding.modbus.stiebeleltron/src/main/java/org/openhab/binding/modbus/stiebeleltron/internal/handler/StiebelEltronHandlerAllWpm.java:
```diff
@@ -0,0 +1,1582 @@
+     * This method starts the operation of this handler Connect to the slave bridge
+     * Start the periodic polling1
+     */
+    @SuppressWarnings("null")
```
done

GerdZanker reviewed (approved):
I tested the add-on based on latest commit and I can confirm that binding and new thing is working as I expected. Logging on trace level gives enough details for analysis & debugging.

Nice work - thank you @tb4jc
